src/platform/audio/alaudiomanager.cpp

```diff
@@ -0,0 +1,1587 @@
+        Get_Info_For_Audio_Event(request->m_event.object);
+    }
+
+    if (request->m_event.object->Get_Event_Info()->Get_Visibility() == VISIBILITY_WORLD) {
+        return m_soundManager->Can_Play_Now(request->m_event.object);
+    }
+
+    return true;
+}
+
+/**
+ * Plays an audio event.
+ */
+void ALAudioManager::Play_Audio_Event(AudioEventRTS *event)
+{
+    const AudioEventInfo *aud_info = event->Get_Event_Info();
+
+    if (aud_info == nullptr) {
+        return;
+    }
+
+    uintptr_t kill_handle = event->Get_Handle_To_Kill();
+    ALuint source_handle = 0;
+    Utf8String name = event->Get_File_Name();
+    PlayingAudio *pa = new PlayingAudio;
+    Init_Playing_Audio(pa);
+    pa->openal.stopped = false;
+
+    switch (aud_info->Get_Event_Type()) {
+        case EVENT_SPEECH:
+            if (event->Should_Play_Locally()) {
+                Stop_All_Speech();
+            }
+            // Fallthrough
+        case EVENT_MUSIC: {
+            float volume = 1.0f;
+
+            if (aud_info->Get_Event_Type() == EVENT_SPEECH) {
+                volume = m_speechVolume;
+            } else {
+                volume = m_musicVolume;
+            }
+
+            volume *= event->Get_Volume();
+            bool killed = false;
+
+            if (kill_handle != 0) {
+                for (auto it = m_streamList.begin(); it != m_streamList.end(); ++it) {
+                    if (*it != nullptr && (*it)->openal.audio_event != nullptr
+                        && (*it)->openal.audio_event->Get_Playing_Handle() == kill_handle) {
+                        Release_Playing_Audio(*it);
+                        m_streamList.erase(it);
+                        killed = true;
+                        break;
+                    }
+                }
+            }
+
+            if (kill_handle != 0 && !killed) {
+                source_handle = 0;
+            } else {
+                alGenSources(1, &source_handle);
+            }
+
+            pa->openal.audio_event = event;
+            pa->openal.source = source_handle;
+            pa->openal.playing_type = PAT_STREAM;
+            break;
+        }
+        case EVENT_SOUND:
+            if (event->Is_Positional_Audio()) {
+                bool killed = false;
+
+                if (kill_handle != 0) {
+                    for (auto it = m_positionalAudioList.begin(); it != m_positionalAudioList.end(); ++it) {
+                        if ((*it)->openal.audio_event->Get_Playing_Handle() == kill_handle) {
+                            Release_Playing_Audio(*it);
+                            m_positionalAudioList.erase(it);
+                            killed = true;
+                            break;
+                        }
+                    }
+                }
+
+                if (kill_handle != 0 && !killed) {
+                    source_handle = 0;
+                } else {
+                    alGenSources(1, &source_handle);
```

Comment:
This looks different to original implementation. Is intended?

Reply:
In which regard does this look different? It should be what MilesAudioManager does in L1747

Reply:
```c++
                if (kill_handle != 0 && !killed) {
                    sample_handle = nullptr;
                } else {
                    sample_handle = Get_First_3D_Sample();

                    if (sample_handle == nullptr && Kill_Lowest_Priority_Sound_Immediately(event)) {
                        sample_handle = Get_First_3D_Sample();
                    }
                }
```


Same at other occurrences of `alGenSources`.